### PR TITLE
🚀 Performance enhancements

### LINKED ISSUE #51: Cache flush command fails when no matching page is found
The flush command returns a 500 error when there is no matching page found.

### LINKED ISSUE #158: Gzip/Brotli response before caching
Gzipping data is an expensive process, the results of the gzip should be cached rather than caching the raw data and having to perform the gzip for every request.

Web v1.10.x

### LINKED ISSUE #173: redirectPort: listener should trigger a 302 redirect
Currently it [sends a 301](https://github.com/dadi/web/blob/49ff8c1b79a83436dfa3da670205c670f25aefc6/dadi/lib/api/index.js#L267), I suggest changing this to a 302 so the redirect is not permanently cached - which is a pain when debugging etc.

N.B. research says search engines are indifferent to redirect type.

### LINKED ISSUE #174: Introduce brotli compression
Brotli is an alternative compression that provides, in real world testing, around 20% smaller payload than gzip.
useful package: https://www.npmjs.com/package/brotli
useful resources: 
https://css-tricks.com/brotli-static-compression/
https://engineering.linkedin.com/blog/2017/05/boosting-site-speed-using-brotli-compression
https://hacks.mozilla.org/2015/11/better-than-gzip-compression-with-brotli/

Also applicable to CDN.

A potentially significant performance increase with a minor technical overhead.

Thoughts please.

### LINKED ISSUE #175: Remove 'sever' response header
Currently we're advertising in the response header `server: DADI (Web)`, this is bad security practice, especially for our clients. Any in-progress security bugs could be used as exploits.

Remove instances of `res.setHeader('Server', config.get('server.name'))`

### LINKED ISSUE #212: Default workspace config error
If starting the app with a very minimal config file with `paths` not defined:

```json
{
  "server": {
    "host": "127.0.0.1",
    "port": 3001
  },
  "cluster": false
}
```

The following error is thrown:

```
> node ./start.js

path.js:7
    throw new TypeError('Path must be a string. Received ' + inspect(path));
    ^

TypeError: Path must be a string. Received undefined
    at assertPath (path.js:7:11)
    at Object.resolve (path.js:1146:7)
    at Server.loadPaths (/web/dadi/lib/index.js:376:28)```

### PR DESCRIPTION
**Pull request by abovebored**
This contains most of the improvements from the on hold #167.

Closes #174.
Closes #175.
Closes #158.
Closes #168.
Closes #173.
Closes #51.
Closes #212.

# Features

* Rolled our own  internal middleware (see `/lib/view/public.js`) for serving from the public folder - removed need for Express modules [serve-static](https://github.com/expressjs/serve-static) and [serve-favicon](https://github.com/expressjs/serve-favicon).
* Changed & deprecated `config.headers.useGzipCompression` to `config.headers.useCompression`
* Added an internal helper called `help.canCompress` which determines if caching is enabled on the system and tests if the client can accept a compressed output
* Implement more efficient compression engine ([brotli](https://www.npmjs.com/package/brotli), by way of [iltorb](https://www.npmjs.com/package/iltorb), fallbacks to Node's `zlib` for `gzip` if not available)
* For clarify, moved the helpers previously name `sendBackJSON`, `sendBackHTML` into `view/send.js`
* Removed the unused helper `sendBackJSONP`
* Added new compression engine to the response, removing need for Express's [Compress](https://www.npmjs.com/package/compression) middleware
* Static assets obey compression settings via `config.headers.useCompression`
* Compressed responses are now cached, rather than being cached on each request
* Assets that have been compressed are cached, rather than being compressed on each request
* Public folder assets are only compressed if it will save space (checked using [compressible](https://www.npmjs.com/package/compressible))
* Removed outdated `media` path.
* Enable page caching by default if `caching` is specified in the config.
* Refactor of cache flush under `api/flush`. Added corresponding error page when method is not `POST`.
* Fix `sinon` at version 2 (3 not compatible with `sinon-test` yet)
* Added `npm run format` to run for [standard](https://www.npmjs.com/package/standard) & [prettier](https://www.npmjs.com/package/prettier)

# To-do

- [x] Add more tests
- [x] Fix Safari "cannot decode raw data" error
- [x] Fix virtualDirectories

### CODE REVIEW

**Comment by eduardoboucas** [comment]
This is awesome! 👏

**Comment by abovebored** [comment]
Because I backtracked the HTTP2 support (#167), I think we should squash this before we merge.

File: dadi/lib/cache/index.js [hunk 1]
```diff
@@ -130,14 +128,23 @@ Cache.prototype.getEndpointMatchingLoadedPaths = function (req) {
...
 
-  if (!endpoint) {
-    endpoint = this.getEndpointMatchingLoadedPaths(req)
-  }
+  return endpoint && endpoint.page && endpoint.page.contentType
+    ? endpoint.page.contentType
+    : false
+}
+
+/**
```
Comment (jimlambie) [inline dadi/lib/cache/index.js]: move comment to next block (L161)

File: dadi/lib/page/index.js [hunk 2]
```diff
@@ -18,6 +18,7 @@ var Page = function (name, schema, hostKey, templateCandidate) {
   this.events = schema.events || []
   this.preloadEvents = schema.preloadEvents || []
   this.requiredDatasources = schema.requiredDatasources || []
+  this.pushManifest = schema.pushManifest || []
```
Comment (jimlambie) [inline dadi/lib/page/index.js]: still req?

File: dadi/lib/index.js [hunk 3]
```diff
@@ -446,8 +386,8 @@ Server.prototype.loadApi = function (options, reload, callback) {
         help.validateRequestMethod(req, res, 'POST') &&
         help.validateRequestCredentials(req, res)
       ) {
-        return help.clearCache(req, err => {
-          help.sendBackJSON(200, res, next)(err, {
+        return help.clearCache(req, this.cacheLayer, err => {
+          Send.json(200, res, next)(err, {
             result: 'success',
             message: 'Succeed to clear'
```
Comment (jimlambie) [inline dadi/lib/index.js]: english?

File: dadi/lib/view/public.js [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+      next,
+      false,
+      headers,
+      shouldCompress,
+      false
+    )
+  }
+}
+
+Public.prototype.openStream = function (
```
Comment (jimlambie) [inline dadi/lib/view/public.js]: use an object argument instead? `{ res: res, file: file }` ?

File: scripts/init-web.js [hunk 5]
```diff
@@ -1,23 +1,27 @@
 #! /usr/bin/env node
 
-var fs = require('fs')
-var path = require('path')
+var fs = require("fs")
```
Comment (jimlambie) [inline scripts/init-web.js]: single quotes preferred 😬

**Comment by adamkdean** [comment]
This is an awesome effort @abovebored 🚀 🌔

**Pull request merged**